File: chazm-model/src/main/java/runtimemodels/chazm/model/id/StringId.java
package runtimemodels.chazm.model.id;

import com.google.inject.assistedinject.Assisted;
import runtimemodels.chazm.model.validation.Checks;

import javax.inject.Inject;
import javax.validation.constraints.NotNull;
import java.util.Objects;

class StringId<T> extends AbstractId<T> {

    private static final long serialVersionUID = 522905742372399827L;
    private final String id;
    private transient Integer hashCode = null;
    private transient String toString = null;

    @Inject
    StringId(@NotNull @Assisted final Class<T> type, @NotNull @Assisted final String id) {
        super(type);
        Checks.checkNotNull(id, "id"); //$NON-NLS-1$
        this.id = id;
    }

    @Override
    public boolean equals(final Object object) {
        if (object instanceof StringId) {
            final StringId<?> other = (StringId<?>) object;
            return super.equals(other) && getId().equals(other.getId());
        }
        return false;
    }

    @Override
    public int hashCode() {
        if (hashCode == null) {
            hashCode = Objects.hash(super.hashCode(), getId());
        }
        return hashCode;
    }

    @Override
    public String toString() {
        if (toString == null) {
            toString = super.toString() + ":" + getId();
        }
        return toString;
    }

    public String getId() {
        return this.id;
    }
}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/id/StringId.kt
package runtimemodels.chazm.model.id

import com.google.inject.assistedinject.Assisted
import java.util.*
import javax.inject.Inject

@Deprecated("")
internal class StringId<T> @Inject constructor(
    @Assisted type: Class<T>,
    @param:Assisted val id: String
) : AbstractId<T>(type) {
    override fun equals(other: Any?): Boolean {
        if (other is StringId<*>) {
            return super.equals(other) && id == other.id
        }
        return false
    }

    override fun hashCode(): Int = Objects.hash(super.hashCode(), id)

    override fun toString(): String = "${super.toString()}:$id"

    companion object {
        private const val serialVersionUID = 522905742372399827L
    }
}
